onnxruntime/core/optimizer/compute_optimizer/compute_optimizer.h

```diff
@@ -0,0 +1,206 @@
+ *
+ * In terms of file dependency, compute_optimizer.h/cc reference structs and utilities defined in
+ * passthrough_actors.h/cc.
+ */
+class ComputeOptimizer : public GraphTransformer {
+ public:
+  ComputeOptimizer(const InlinedHashSet<std::string_view>& compatible_execution_providers = {}) noexcept
+      : GraphTransformer("ComputeOptimizer", compatible_execution_providers) {}
+
+  Status ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const override;
+
+ private:
+  std::optional<SliceInfo> IsSupportedGatherND(Graph& graph, Node& node, const logging::Logger& logger) const;
+  std::optional<SliceInfo> IsSupportedGather(Graph& graph, Node& node, const logging::Logger& logger) const;
+};
+
+namespace optimizer {
+namespace compute_optimizer {
+
+using OPSET_VERSION_LIST = std::initializer_list<ONNX_NAMESPACE::OperatorSetVersion>;
+const OPSET_VERSION_LIST opset_1{1};
+const OPSET_VERSION_LIST opset_13_1{13, 1};
+const OPSET_VERSION_LIST opset_13_9_1{13, 9, 1};
+const OPSET_VERSION_LIST opset_13_11_1{13, 11, 1};
+const OPSET_VERSION_LIST opset_13_9_6_1{13, 9, 6, 1};
+const OPSET_VERSION_LIST opset_14_13_5_1{14, 13, 5, 1};
+const OPSET_VERSION_LIST opset_14_13_7_6_1{14, 13, 7, 6, 1};
+const OPSET_VERSION_LIST opset_13_12_10_7_6_1{13, 12, 10, 7, 6, 1};
+
+/**
+ * @brief Functor to trigger the optimization search for a given slicing node
+ *   (for example Gather/GatherND node).
+ */
+struct SliceOperationReorderHandle {
+  /**
+   * @brief Pass through configuration for specific operator.
+   *
+   * For each operator:
+   * > `input_indices` can be used to explicitly specify the input indices that Slicing op can be passed through.
+   *   This could be helpful if some inputs are not applicable for pass through. If not specified, all inputs
+   *   are considered (but there will be checks to ignore those inputs that are not affected by the slicing axis).
+   * > `actor` will be used to perform the actual pass through, including both pre-check stage and post process
+   *   stage.
+   */
+  struct OpPassThroughConfig {
+    OpPassThroughConfig(const std::vector<int>& input_indices,
+                        std::shared_ptr<OperatorPassThroughActorBase> actor,
+                        const OPSET_VERSION_LIST& opset_list)
+        : input_indices(input_indices), actor(actor), opsets(opset_list) {
+    }
+
+    std::vector<int> input_indices;
+    std::shared_ptr<OperatorPassThroughActorBase> actor;
+    const OPSET_VERSION_LIST& opsets;
+  };
+
+  static std::string GetFullQualifiedOpName(const std::string& op_type, const std::string& domain) {
+    return domain + "::" + op_type;
+  }
+
+  static std::unordered_map<std::string, OpPassThroughConfig>& GetOpPassThroughConfigMap() {
+    static std::unordered_map<std::string, OpPassThroughConfig> allowed_passthrough_ops;
+    static std::once_flag allowed_ops_init;
+    std::call_once(allowed_ops_init, []() {
```

Comment:
No need to use calll_once. put the body of the this lamda expr in a function, then write:

    static std::unordered_map<std::string, OpPassThroughConfig> allowed_passthrough_ops = xxx();